refactor: major extension classes refactoring

friday opened the pull request:
Fixes #1044
This PR should also help for #1018 (need a common API for the UI and cmd installation) and #1063.

Creates a new `ExtensionController` class (not to be confused with the old `ExtensionController` class which I just renamed and pushed before this PR (e3af72a1)

`ExtensionController` is intended to be the single interface to be called by other classes, and to internally handle ExtensionManifest, ExtensionRemote, ExtensionRuntime and ExtensionDb/Record (now renamed to `ExtensionState`) so other classes won't touch those classes won't handle them, and they won't handle each other (with a couple of necessary exceptions for ExtensionManifest).

* `ExtensionDownloader` no longer exists (it is now part of ExtensionController)
* ExtensionRunner has been rewritten to "ExtensionRuntime". ExtensionRuntime handle one single instance and is really just a thin abstraction over Gio.Subprocess with Ulauncher extension specific stderr error handling. `ExtensionController.start()` handles the manifest validation, updating the state and preparing the ENV variables.
* `ExtensionState` (the new ExtensionDb/Record) is integrated into ExtensionController and only accessible from there. It now creates individual json files for each extension (like extension settings) instead of storing all to `extensions.json`

I also opted to rewrite the ExtensionRuntime to use asyncio.subprocess instead of Gio.Subprocess. I think that would be better for the architecture and maintenance, but I opted not to do it because we had some issues before Troy rewrote to use Gio.Subprocess.

friday commented:
@nazarewk 
Happy new year!

I made some much needed refactoring work on the extension classes, and I'm not sure how to handle PRs like this since I don't have the means to verify that I didn't break the nix package. Do you think it would be possible to add something to our [GitHub linting CI workflow](https://github.com/Ulauncher/Ulauncher/blob/v6/.github/workflows/tests.yml) that could continuously verify that the package can still be installed?

nazarewk commented:
Hello, I had a lot going on personally and didn't have a chance to move the topic forward.

While updating my system I just noticed the `mypy` failing and pinned the version to latest working commit as fixing it required some research on generating proper stubs.

friday commented:
> Hello, I had a lot going on personally and didn't have a chance to move the topic forward.
> 
> While updating my system I just noticed the `mypy` failing and pinned the version to latest working commit as fixing it required some research on generating proper stubs.

Ok. I can definitely relate to having to fight the linter tools. There's no hurry here. I will move on with PR in the meanwhile and you can make a new PR when you have time.

And I will rename `.extension-record.json` to `.default-state.json`.

nazarewk commented:
yeah no problem, I'm not using it yet anywhere anyway

nazarewk commented:
I wonder what is the Dockerfile for `ulauncher/build-image:6.0`?

friday commented:
> I wonder what is the Dockerfile for `ulauncher/build-image:6.0`?

This one: https://github.com/Ulauncher/Ulauncher/blob/v6/Dockerfile
I remember I had to revert/rebuild to fix some release (it's also used in the CI), so the published version might not have been the same as the current v6 branch.
If it doesn't work locally, maybe try building with `docker build -t ulauncher/build-image:6.0` and that will be used by the `./ul dev-container` script.

nazarewk commented:
I figured out the error, can you include this patch?

```patch
Subject: [PATCH] nix: fix build
---
Index: nix/default.nix
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/nix/default.nix b/nix/default.nix
--- a/nix/default.nix	(revision 07de958e41a732dc5f11e9a79fe3e20f00bc02d9)
+++ b/nix/default.nix	(revision 4265f04fbc6f933a0d8bbdc308c98242a540a167)
@@ -62,6 +62,7 @@
   packages.tests.python = pp: (with pp; [
     black
     mock
+    (pygobject-stubs.overridePythonAttrs (old: { PYGOBJECT_STUB_CONFIG = "Gtk3,Gdk3,Soup2"; }))
     pytest
     pytest-mock
   ]);
@@ -126,7 +127,6 @@
         --replace gapplication ${glib}/bin/gapplication
       substituteInPlace \
           ulauncher/modes/extensions/ExtensionController.py \
-        --replace 'from ulauncher.config import PATHS,' 'from ulauncher.config import' \
         --replace '"PYTHONPATH": PATHS.APPLICATION,' '"PYTHONPATH": ":".join(sys.path),'
 
       substituteInPlace \
```

nazarewk commented on nix/default.nix:
```diff
@@ -125,7 +125,7 @@ let
       substituteInPlace \
-          ulauncher/modes/extensions/ExtensionRunner.py \
+          ulauncher/modes/extensions/ExtensionController.py \
         --replace 'from ulauncher.config import PATHS,' 'from ulauncher.config import' \
```
```suggestion
```

nazarewk replied:
also this:

```patch
Subject: [PATCH] nix: fix build
---
Index: nix/default.nix
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/nix/default.nix b/nix/default.nix
--- a/nix/default.nix	(revision 07de958e41a732dc5f11e9a79fe3e20f00bc02d9)
+++ b/nix/default.nix	(revision 4265f04fbc6f933a0d8bbdc308c98242a540a167)
@@ -62,6 +62,7 @@
   packages.tests.python = pp: (with pp; [
     black
     mock
+    (pygobject-stubs.overridePythonAttrs (old: { PYGOBJECT_STUB_CONFIG = "Gtk3,Gdk3,Soup2"; }))
     pytest
     pytest-mock
   ]);
```

nazarewk reviewed (changes requested):
Changes joined together, but I'm not sure if that Github action would work:
```patch
Subject: [PATCH] nix: fix build
---
Index: nix/default.nix
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/nix/default.nix b/nix/default.nix
--- a/nix/default.nix	(revision 07de958e41a732dc5f11e9a79fe3e20f00bc02d9)
+++ b/nix/default.nix	(date 1704284789757)
@@ -62,6 +62,7 @@
   packages.tests.python = pp: (with pp; [
     black
     mock
+    (pygobject-stubs.overridePythonAttrs (old: { PYGOBJECT_STUB_CONFIG = "Gtk3,Gdk3,Soup2"; }))
     pytest
     pytest-mock
   ]);
@@ -126,7 +127,6 @@
         --replace gapplication ${glib}/bin/gapplication
       substituteInPlace \
           ulauncher/modes/extensions/ExtensionController.py \
-        --replace 'from ulauncher.config import PATHS,' 'from ulauncher.config import' \
         --replace '"PYTHONPATH": PATHS.APPLICATION,' '"PYTHONPATH": ":".join(sys.path),'
 
       substituteInPlace \
Index: .github/workflows/tests.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/workflows/tests.yml b/.github/workflows/tests.yml
--- a/.github/workflows/tests.yml	(revision 07de958e41a732dc5f11e9a79fe3e20f00bc02d9)
+++ b/.github/workflows/tests.yml	(date 1704285533613)
@@ -20,6 +20,9 @@
         run: DISPLAY=:1 xvfb-run pytest tests
       - name: build preferences
         run: ./setup.py build_prefs --force
+      - uses: cachix/install-nix-action@v22
+      - name: nix build
+        run: nix build --out-link nix/result --show-trace --print-build-logs .#default
       - name: build docs
         run: |
           cd docs
```

friday commented:
Thank you! I split it up in two commits so I could test the build would fail, then work.

But it seems it's failing now because the docker image doesn't have curl. I'm building a new one now trying to fix that.

nazarewk commented:
Yeah, I'm not familiar with GitHub Actions at all, I guess those 3 lines would work in whatever default configuration is provisioned by GHA.

friday commented:
I moved the `(pygobject-stubs.overridePythonAttrs (old: { PYGOBJECT_STUB_CONFIG = "Gtk3,Gdk3,Soup2"; }))` part to a sepatate commit to v6, amended the new install-nix-action workflow (for now) and squashed the other fix to the commit that caused it.

So it should work now, minus the new workflow. Will try that in a separate branch.
